feat: allow chart deploy overrides (#2403)
defenseunicorns/zarf: adr/0023-chart-override.md

Reviewer: needs a demonstration of usage in this ADR of both a minimal `zarf.yaml` to demonstrate and the minimal `zarf` commands to replicate.

Author: I have included an example in the ADR for a minimal `zarf.yaml` and command line flags to use it.

# 23. Simplifying Helm Chart Value Overrides

Date: 2024-03-25

## Status

Proposed

## Context

> The process of deploying applications with Helm charts in Kubernetes environments often necessitates the customization of chart values to align with specific operational or environmental requirements. The current method for customizing these values—either through manual edits or `###`. A more streamlined approach would greatly enhance the deployment experience by offering both flexibility and reliability.

Reviewer: ```suggestion
The process of deploying applications with Helm charts in Kubernetes environments often necessitates the customization of chart values to align with specific operational or environmental requirements. The current method for customizing these values—either through manual edits or `###ZARF_VAR_XYZ###`. A more streamlined approach would greatly enhance the deployment experience by offering both flexibility and reliability.
```

## Decision

To address this issue, we propose the introduction of a feature designed to simplify the process of overriding chart values at the time of deployment. This feature would allow users to easily specify overrides for any chart values directly via command-line arguments, eliminating the need to alter the chart's default values file or manage multiple command-line arguments for each override.

Key aspects of the proposed implementation include:
- Use existing `--set`  flags to specify overrides for chart values.

> - The ability to list all overrides in a structured and easily understandable format within the ZarfConfig file.

Reviewer: ```suggestion
- The ability to list all overrides in a structured and easily understandable format within `zarf.yaml`.
```

- Ensuring that during deployment, these specified overrides take precedence over the chart's default values, thus facilitating customized deployments without necessitating permanent modifications to the chart.

## Consequences

Adopting this feature would lead to several key improvements:

> - **Streamlined Configuration Process**: Centralizing overrides in a single, unified file significantly simplifies the management of configuration settings, aligning more closely with standard Helm practices and reducing the reliance on extensive command-line arguments.

Author: I fixed this. Thanks

> - **Streamlined Configuration Process**: Centralizing overrides in a single, unified file significantly simplifies the management of configuration settings, aligning more closely with standard Helm practices and reducing the reliance on extensive custom `###` templating.

Reviewer: ```suggestion
- **Streamlined Configuration Process**: Centralizing overrides in a single, unified file significantly simplifies the management of configuration settings, aligning more closely with standard Helm practices and reducing the reliance on extensive custom `###ZARF_VAR_XYZ###` templating.
```

## Decision

To address this issue, we propose the introduction of a feature designed to simplify the process of overriding chart values at the time of deployment. This feature would allow users to easily specify overrides for any chart values directly via command-line arguments, eliminating the need to alter the chart's default values file or manage multiple command-line arguments for each override.

Key aspects of the proposed implementation include:
- Use existing `--set`  flags to specify overrides for chart values.
- The ability to list all overrides in a structured and easily understandable format within `zarf.yaml`.
- Ensuring that during deployment, these specified overrides take precedence over the chart's default values, thus facilitating customized deployments without necessitating permanent modifications to the chart.

## Consequences

Adopting this feature would lead to several key improvements:

> - **Streamlined Configuration Process**: Centralizing overrides in a single, unified file significantly simplifies the management of configuration settings, aligning more closely with standard Helm practices and reducing the reliance on extensive custom `###ZARF_VAR_XYZ###` templating.

Reviewer 2: I wouldn't say this centralizes overrides, in fact, now overrides can happen in multiple files (values.yaml and zarf.yaml). 
Maybe some thing more like
```
- **Streamlined Configuration Process**: Allowing helm values overrides in the zarf package schema simplifies the user experience by reducing the reliance on extensive custom `###ZARF_VAR_XYZ###` templating and aligning more closely with standard Helm practices
```

Author: I took care of it.
